### ISSUE TITLE
Macros, strings, and escapes issue.

### ISSUE THREAD

**Issue by Hinaichigo** [OpenDreamProject/OpenDream#700]
Trying to compile `vgstation13-coders/vgstation13`, the following produces the error:
`Error at code\modules\paperwork\pen.dm:23:2: Expected ']'`
```
13	#define REG_NOTBB "\[^\\\[\]+"    // [^\]]+
14	
15	// This WAS a macro, but BYOND a shit.
16	/proc/REG_BBTAG(x)
17		return "\\\[[x]\\\]"
18	
19	// [x]blah[/x]
20	/proc/REG_BETWEEN_BBTAG(x)
21		return "[REG_BBTAG(x)]([REG_NOTBB])[REG_BBTAG("/[x]")]"
22	
23	/datum/speech_filter_action/bbcode/Run(var/text, var/mob/user, var/atom/movable/P)
24		return
```
Substituting everything into line 21, it becomes:
`return "["\\\[[x]\\\]"](["\[^\\\[\]+"])["\\\[["/[x]"]\\\]"]"`

The brackets are not balanced in that string, but it seems that they are brought to parity at runtime.

**Comment by wixoaGit** [comment]
The brackets aren't unbalanced. `\[` in a string escapes that bracket and isn't used in string interpolation.

**Comment by out-of-phaze** [comment]
It looks more like your macro is missing some escapes: `#define REG_NOTBB "\[^\\\[\]+"    // [^\]]+` tries to interpolate `[^\]`.

**Comment by out-of-phaze** [comment]
Ah, looks like I misread. That's a comment. `"\[^\\\[\]+"` is the full macro.
Can you post what the preprocessor output is?

**Comment by Hinaichigo** [comment]
How do I obtain that?

**Comment by wixoaGit** [comment]
> How do I obtain that?

Add the `--dump-preprocessor` argument when running the compiler. This will create a new file next to the DME called `preprocessor_dump.dm`.

**Comment by Hinaichigo** [comment]
Thanks. The file has over 500k lines. Is there a certain section that I should post?

**Comment by wixoaGit** [comment]
The section of code you included in the top of the issue

**Comment by Hinaichigo** [comment]
Oh I just realized what was meant:

```
#define REG_NOTBB "\[^\\\[\]+"    // [^\]]+

// This WAS a macro, but BYOND a shit.
/proc/REG_BBTAG(x)
	return "\\\[[x]\\\]"

// [x]blah[/x]
/proc/REG_BETWEEN_BBTAG(x)
	return "[REG_BBTAG(x)]([REG_NOTBB])[REG_BBTAG("/[x]")]"
```

**Comment by wixoaGit** [comment]
Did that come from the preprocessor dump file? Preprocessor commands like `#define` shouldn't show up in there.

**Comment by Hinaichigo** [comment]
Sorry, I edited it to the correct info.

**Comment by wixoaGit** [comment]
I got this down to a smaller test case:
`"[ "\[" ]"`

I suspect that the `\[` inside the interpolation is causing the `bracketNesting` counter to get confused when parsing the string. This would cause it to expect a second `]` to end an interpolation that never existed.

**Comment by out-of-phaze** [comment]
> I got this down to a smaller test case: `"[ "\[" ]"`
> 
> I suspect that the `\[` inside the interpolation is causing the `bracketNesting` counter to get confused when parsing the string. This would cause it to expect a second `]` to end an interpolation that never existed.

That appears to be a different issue, since I was able to fix that one locally by changing the priority of escapes vs brackets in the preprocessor lexer, but the main issue remains.

**Comment by Hinaichigo** [comment]
I want to try to fix this, but just for an initial sanity check and base to start from:

It should become:
`"["\\\[[x]\\\]"](["\[^\\\[\]+"])["\\\[["/[x]"]\\\]"]"`
(If I put that into line 21, the same compile error occurs, so I think that's safe to say there's no issues with using the macro in the preprocessor or evaluating the `REG_BBTAG()` proc?)

And each `\[` and `\]` are supposed to just be treated as the literal characters `[` and `]` and not used for increasing `bracketNesting` or related considerations or anything?
So as a test I replaced `\[` with `a` and `\]` with `b`:
`"["\\a[x]\\b"](["a^\\ab+"])["\\a["/[x]"]\\b"]"`
And that compiles without erroring.
So it does seem that the problem is indeed caused by the `\[` and `\]`

But in `"["\\a[x]\\b"](["a^\\ab+"])["\\a["/[x]"]\\b"]"`...
`\\` is just meant to be interpreted as the literal `\` character right?
So if we replace `\\` with `c` (as the first step, before then replacing `\[` with `a` and `\]` with `b`) it becomes:
`"["ca[x]cb"](["a^cab+"])["ca["/[x]"]cb"]"` 
In that case I think it makes sense the highest value `bracketNesting` should reach is 3. And if I log the max value of `bracketNesting` it is 3...

However, if I instead plug in `"["\\\[[x]\\\]"](["\[^\\\[\]+"])["\\\[["/[x]"]\\\]"]"` in place of line 21's `"[REG_BBTAG(x)]([REG_NOTBB])[REG_BBTAG("/[x]")]"`, and log the maximum value of `bracketNesting`, it is 5.

So it does seems that there are two issues, since there is different behavior when slotting in the expected values explicitly versus letting the compiler obtain their values naturally (let's call this issue B), confounded with the issue of the supposedly unbalanced brackets (issue A) Maybe issue B is related to the way `"` quotes are handled in these situations, where there's a disparity between BYOND and OD?

**Comment by wixoaGit** [comment]
Sorry for the late response. I've looked into this, and it does appear that the issue comes from `bracketNesting` going up when it's escaped in a string inside an interpolated expression.

> For clarity, what is the rationale behind having '\\' be parsed differently inside versus outside of brackets here:

It's for (ridiculous) cases like:
```
var/\proper = 1 + 2
world.log << "[\proper]"
```
Where backspaced tokens shouldn't be treated as escape codes inside an interpolated value.

Trying to parse edge cases like this in a string could get very messy very quickly, so the best solution would probably be to keep string tokens and interpolated value tokens separate when parsing. The preprocessor's lexer already keeps the tokens separate, but the DM lexer [stitches them back together](https://github.com/OpenDreamProject/OpenDream/blob/f7f90cecdf154eff1eb6e5cccefe3d4e3e3529a8/DMCompiler/Compiler/DM/DMLexer.cs#L259). It should be a matter of not doing that and having the parser handle it.

**Comment by amylizzle** [comment]
Goon blocker now too. Choking on 
https://github.com/goonstation/goonstation/blob/94b73ac17c3801312b2f46174b504edee150834c/code/mob/living/critter/flock/flockbit.dm#L39
and
https://github.com/goonstation/goonstation/blob/94b73ac17c3801312b2f46174b504edee150834c/code/obj/machinery/phone/phone.dm#L385

**Comment by wixoaGit** [comment]
Fixed by #1555
